hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java

```diff
@@ -312,101 +224,101 @@ protected ClientInvocationFuture invoke(ClientMessage req, Data keyData, int com
     }
 
     protected <T> ICompletableFuture<T> getAndRemoveAsyncInternal(K key, boolean withCompletionEvent, boolean async) {
-        final long start = System.nanoTime();
+        final long startNanos = nowInNanosOrDefault();
         ensureOpen();
         validateNotNull(key);
-        CacheProxyUtil.validateConfiguredTypes(cacheConfig, key);
+        validateConfiguredTypes(cacheConfig, key);
+
         final Data keyData = toData(key);
+
         final int completionId = withCompletionEvent ? nextCompletionId() : -1;
         ClientMessage request = CacheGetAndRemoveCodec.encodeRequest(nameWithPrefix, keyData, completionId);
-        ClientInvocationFuture future;
-        try {
-            future = invoke(request, keyData, completionId);
-            invalidateNearCache(keyData);
-        } catch (Exception e) {
-            throw rethrow(e);
-        }
-        ClientDelegatingFuture<T> delegatingFuture =
-                new ClientDelegatingFuture<T>(future, clientContext.getSerializationService(), GET_AND_REMOVE_RESPONSE_DECODER);
-        if (async && statisticsEnabled) {
-            delegatingFuture.andThenInternal(new ExecutionCallback<T>() {
+        ClientInvocationFuture future = invoke(request, keyData, completionId);
+        ClientDelegatingFuture<T> delegatingFuture
+                = new ClientDelegatingFuture<T>(future, serializationService, GET_AND_REMOVE_RESPONSE_DECODER);
+        if (async) {
+            ExecutionCallback<T> callback = !statisticsEnabled ? null : new ExecutionCallback<T>() {
                 public void onResponse(T response) {
-                    handleStatisticsOnRemove(true, start, response);
+                    statsHandler.onRemove(true, startNanos, response);
                 }
 
                 public void onFailure(Throwable t) {
 
                 }
-            }, true);
+            };
+
+            onGetAndRemoveAsyncInternal(keyData, delegatingFuture, callback);
         }
         return delegatingFuture;
     }
 
-    protected <T> ICompletableFuture<T> removeAsyncInternal(K key, V oldValue, boolean hasOldValue, boolean withCompletionEvent,
-                                                            boolean async) {
-        final long start = System.nanoTime();
+    protected <T> void onGetAndRemoveAsyncInternal(Data keyData, ClientDelegatingFuture<T> delegatingFuture,
+                                                   ExecutionCallback<T> callback) {
+        addCallback(delegatingFuture, callback);
+    }
+
+    protected Object removeAsyncInternal(K key, V oldValue, boolean hasOldValue, boolean withCompletionEvent,
```

Comment:
I think it's worth to split up `removeAsyncInternal()` into a sync and async method, like we have it for `get()`. This `boolean` parameter and `Object` as return type just causes warnings and readability issues (three boolean parameters, very easy to get confused).

But this should not be done in this PR.